File: queue/php/src/app/Queue.php
<?php
declare(strict_types=1);

class Queue
{
    private int $capacity = 8;

    /**
     * @var int[]
     */
    private array $values = [];

    /**
     * Allocates enough memory for a required capacity.
     */
    public function allocate(int $capacity): void
    {
        $this->capacity = $capacity;
    }

    /**
     * Returns the current capacity.
     */
    public function getCapacity(): int
    {
        return $this->capacity;
    }

    /**
     * Clear queue.
     */
    public function clear(): void
    {
        $this->values = [];
    }

    /**
     * Returns a shallow copy of the queue.
     */
    public function copy(): Queue
    {
        return clone $this;
    }

    /**
     * Returns whether the queue is empty.
     */
    public function isEmpty(): bool
    {
        return count($this->values) < 1;
    }

    /**
     * Returns the value at the front of the queue.
     */
    public function peek(): int
    {
        return $this->values[0];
    }

    /**
     * Removes and returns the value at the front of the queue.
     */
    public function pop(): int
    {
        return array_shift($this->values);
    }

    /**
     * Pushes values into the queue.
     */
    public function push(int $value): void
    {
        if ($this->capacity === count($this->values)) {
            throw new InvalidArgumentException(
                'Not enough memory. Capacity: ' . $this->capacity
            );
        }

        $this->values[] = $value;
    }

    /**
     * Converts the queue to an array.
     */
    public function toArray(): array
    {
        return $this->values;
    }
}

File: queue/go/app/queue.go
package main

import (
	"errors"
	"fmt"
)

type Queue struct {
	capacity int
	values   []int
}

func (q *Queue) allocate(capacity int) {
	q.capacity = capacity
}

func (q *Queue) getCapacity() int {
	return q.capacity
}

func (q *Queue) clear() {
	q.values = nil
}

func (q Queue) copy() Queue {
	clone := q

	return clone
}

func (q *Queue) isEmpty() bool {
	return len(q.values) < 1
}

func (q *Queue) push(value int) {
	if len(q.values) == q.capacity {
		errors.Unwrap(fmt.Errorf("Not enough memory. Capacity: %d", q.capacity))
	}

	q.values = append(q.values, value)
}

func (q *Queue) peek() int {
	return q.values[0]
}

func (q *Queue) pop() int {
	var value int

	value, q.values = q.values[0], q.values[1:]

	return value
}
